cellxgene_schema_cli/tests/test_schema_compliance.py

```diff
@@ -0,0 +1,1006 @@
+        """
+
+        self.validator.adata.uns["default_embedding"] = "X_other"
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: 'X_other' in 'uns['default_embedding']' is not valid, "
+                "it must be a key of 'adata.obsm'."
+            ],
+        )
+
+    def test_X_approximate_distribution_is_str(self):
+
+        """
+        X_approximate_distribution str. The value MUST be "count" [...] or "normal"
+        """
+
+        self.validator.adata.uns["X_approximate_distribution"] = ["count"]
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: '['count']' in 'uns['X_approximate_distribution']' "
+                "is not valid, it must be a string."
+            ],
+        )
+
+    def test_X_approximate_distribution_is_valid(self):
+
+        """
+        X_approximate_distribution str. The value MUST be "count" [...] or "normal"
+        """
+
+        self.validator.adata.uns["X_approximate_distribution"] = "COUNT"
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: 'COUNT' in 'uns['X_approximate_distribution']' is "
+                "not valid. Allowed terms: ['count', 'normal']."
+            ],
+        )
+
+
+class TestObsm(unittest.TestCase):
+
+    """
+    Fail cases for adata.obsm
+    """
+
+    def setUp(self):
+
+        self.validator = validate.Validator()
+        self.validator.adata = examples.adata.copy()
+
+    def test_obsm_values_ara_numpy(self):
+
+        """
+        values in obsm MUST be a numpy.ndarray
+        """
+
+        self.validator.adata.obsm["X_tsne"] = pd.DataFrame(
+            self.validator.adata.obsm["X_umap"], index=self.validator.adata.obs_names
+        )
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: All embeddings have to be of 'numpy.ndarray' type, "
+                "'adata.obsm['X_tsne']' is <class 'pandas.core.frame.DataFrame'>')"
+            ],
+        )
+
+    def test_obsm_values_at_least_one_X(self):
+
+        """
+        At least one key for the embedding MUST be prefixed with "X_"
+        """
+
+        self.validator.adata.obsm["umap"] = self.validator.adata.obsm["X_umap"]
+        self.validator.adata.uns["default_embedding"] = "umap"
+        del self.validator.adata.obsm["X_umap"]
+        self.validator.validate_adata()
+        self.assertEqual(
+            self.validator.errors,
+            [
+                "ERROR: At least one embedding in 'obsm' has to have a "
+                "key with an 'X_' prefix."
+            ],
+        )
+
+    def test_obsm_shape(self):
+
+        """
+        Curators MUST annotate one or more two-dimensional (m >= 2) embeddings
```

Comment:
embeddings in single-cell data are used for other purposes aside visualization, e.g. PCs from PCA can be used for both visualization and  analyses: the first 2 PCs can used to visualize the data whereas the rest of the PCs can be used to perform some sort of analysis on the data that users who download it may want to have access to.